cpp/src/search/contains_table.cu

```diff
@@ -1,148 +1,403 @@
+ * operating on `cudf::size_type`.
+ */
+template <typename Comparator>
+struct strong_index_comparator_adapter {
+  strong_index_comparator_adapter(Comparator const& comparator) : _comparator{comparator} {}
+
+  template <typename T,
+            typename U,
+            CUDF_ENABLE_IF(is_strong_index_type<T>() && is_strong_index_type<U>())>
+  __device__ constexpr auto operator()(T const lhs_index, U const rhs_index) const noexcept
+  {
+    auto const lhs = static_cast<size_type>(lhs_index);
+    auto const rhs = static_cast<size_type>(rhs_index);
+
+    if constexpr (std::is_same_v<T, U> || std::is_same_v<T, lhs_index_type>) {
+      return _comparator(lhs, rhs);
+    } else {
+      // Here we have T == rhs_index_type.
+      // This is when the indices are provided in wrong order, so we need to switch them back
+      // to the right order to enforce symmetry of the comparator.
+      return _comparator(rhs, lhs);
+    }
+  }
+
+ private:
+  Comparator const _comparator;
+};
+
+/**
+ * @brief Build a row bitmask for the input table.
+ *
+ * The output bitmask will have invalid bits corresponding to the the input rows having nested
+ * nulls and vice versa.
+ *
+ * @param input The input table
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A pair of pointer to the output bitmask and the buffer containing the bitmask
+ */
+std::pair<rmm::device_buffer, bitmask_type const*> build_row_bitmask(table_view const& input,
+                                                                     rmm::cuda_stream_view stream)
+{
+  auto const nullable_columns = get_nullable_columns(input);
+  CUDF_EXPECTS(nullable_columns.size() > 0,
+               "The input table has nulls thus it should have nullable columns.");
+
+  // If there are more than one nullable column, we compute `bitmask_and` of their null masks.
+  // Otherwise, we have only one nullable column and can use its null mask directly.
+  auto row_bitmask           = nullable_columns.size() > 1
+                                 ? cudf::detail::bitmask_and(table_view{nullable_columns}, stream).first
+                                 : rmm::device_buffer{0, stream};
+  auto const row_bitmask_ptr = nullable_columns.size() > 1
+                                 ? static_cast<bitmask_type const*>(row_bitmask.data())
+                                 : nullable_columns.front().null_mask();
+  return std::pair(std::move(row_bitmask), row_bitmask_ptr);
+}
+
+/**
+ * @brief Invoke an `operator()` template with a row equality comparator based on the specified
+ * `compare_nans` parameter.
+ *
+ * @param compare_nans The flag to specify whether NaNs should be compared equal or not
+ * @param func The input functor to invoke
+ */
+template <typename Func>
+void dispatch_nan_comparator(nan_equality compare_nans, Func&& func)
+{
+  if (compare_nans == nan_equality::ALL_EQUAL) {
+    using nan_equal_comparator =

```

Comment:
Perhaps these `using` declarations should be collected at the top of the file with the others, and instead of making it this much shorter just drop all the namespacing? e.g. `nan_equal_physical_equality_comparator` instead of `nan_equal_comparator`? That way it's clear what object is referred to without needing the aliasing logic right next to the code.

Reply:
This was suggested by @bdice: https://github.com/rapidsai/cudf/pull/11118#discussion_r904283235